src/libusermode/userhook.cpp

```diff
@@ -0,0 +1,821 @@
+                {
+                    if (make_trap(vmi, drakvuf, info, &target, exec_func))
+                        target.state = HOOK_OK;
+                }
+            }
+            else
+            {
+                target.state = HOOK_FAILED;
+            }
+
+            PRINT_DEBUG("[USERHOOK] Hook %s (vaddr = 0x%llx, dll_base = 0x%llx, result = %s)\n",
+                target.target_name.c_str(),
+                (unsigned long long)exec_func,
+                (unsigned long long)dll_meta->real_dll_base,
+                target.state == HOOK_OK ? "OK" : "FAIL");
+        }
+    }
+
+    drakvuf_release_vmi(drakvuf);
+    PRINT_DEBUG("[USERHOOK] Done, flag DLL as hooked\n");
+    dll_meta->is_hooked = true;
+
+    for (auto& reg : plugin->plugins) {
+        reg.post_cb(drakvuf, dll_meta, reg.extra);
+    }
+
+    return VMI_EVENT_RESPONSE_NONE;
+}
+
+/**
+ * This is used in order to observe when SysWOW64 process is loading a new DLL.
+ * If the DLL is interesting, we perform further investigation and try to equip user mode hooks.
+ */
+static event_response_t protect_virtual_memory_hook_cb(drakvuf_t drakvuf, drakvuf_trap_info_t * info)
+{
+    // IN HANDLE ProcessHandle
+    uint64_t process_handle = drakvuf_get_function_argument(drakvuf, info, 1);
+    // IN OUT PVOID *BaseAddress
+    addr_t base_address_ptr = drakvuf_get_function_argument(drakvuf, info, 2);
+
+    if (process_handle != ~0ULL)
+        return VMI_EVENT_RESPONSE_NONE;
+
+    auto plugin = get_trap_plugin<userhook>(info);
+    if (!plugin)
+        return VMI_EVENT_RESPONSE_NONE;
+
+    dll_t * dll_meta = get_pending_dll(drakvuf, info, plugin);
+
+    if (!dll_meta)
+    {
+        addr_t base_address;
+
+        access_context_t ctx =
+        {
+            .translate_mechanism = VMI_TM_PROCESS_DTB,
+            .dtb = info->regs->cr3,
+            .addr = base_address_ptr
+        };
+
+        vmi_instance_t vmi = drakvuf_lock_and_get_vmi(drakvuf);
+        bool success = (VMI_SUCCESS == vmi_read_addr(vmi, &ctx, &base_address));
+        drakvuf_release_vmi(drakvuf);
+
+        if (!success)
+            return VMI_EVENT_RESPONSE_NONE;
+
+        dll_meta = create_dll_meta(drakvuf, info, plugin, base_address);
+    }
+
+    if (dll_meta)
+        return perform_hooking(drakvuf, info, plugin, dll_meta);
+
+    return VMI_EVENT_RESPONSE_NONE;
+}
+
+/**
+ * This is used in order to observe when 64 bit process is loading a new DLL.
+ * If the DLL is interesting, we perform further investigation and try to equip user mode hooks.
+ */
+static event_response_t map_view_of_section_ret_cb(drakvuf_t drakvuf, drakvuf_trap_info_t * info)
+{
+    auto data = get_trap_params<userhook, map_view_of_section_result_t<userhook>>(info);
+    userhook* plugin = data->plugin();
+    if (!data || !plugin)
+    {
+        PRINT_DEBUG("[USERHOOK] map_view_of_section_ret_cb invalid trap params!\n");
+        drakvuf_remove_trap(drakvuf, info->trap, nullptr);
```

Comment:
If you remove the trap here, how does the trap structure gets freed?

Reply:
This is some pattern that repeats within few other plugins using `pluginex`:

https://github.com/tklengyel/drakvuf/blob/2369abd980c29aceaccd855c6c2d6ee8fd486683/src/plugins/procmon/procmon.cpp#L280
https://github.com/tklengyel/drakvuf/blob/893f95ac98b065eedfe378dc2f53c50b6645369b/src/plugins/wmimon/wmimon.cpp#L463

I wonder what really should be done in such a situation. If there is no plugin/no trap data then basically it means that there is some bug or the same breakpoint was fired for different plugin and different purpose(?).

I bet we should just log such a situation and skip further execution but don't actually remove the trap?

Reply:
Those plugins are probably not managing traps properly either. Every time a trap is allocated dynamically it needs to be registered by the plugin internally in some fashion so that if the plugin has to terminate in the middle it knows exactly where the traps are and can clear them up. Just removing the trap from drakvuf doesn't free it's memory. There is no garbage collection. So while in this case this is an error path, if the trap just gets removed but not freed you end up leaking memory. That needs to be fixed, especially since this is a library component now, not just a plugin.

Reply:
Ok, for now I've just removed this line:
```
drakvuf_remove_trap(drakvuf, info->trap, nullptr);
```
and in such a case (`!plugin || !data`) a debug message will be logged, but DRAKVUF will continue its operation. I wonder whether to add `throw -1` in such cases or not.